lib/web_ui/lib/src/engine/platform_dispatcher.dart

```diff
@@ -60,6 +63,56 @@ class HighContrastSupport {
   }
 }
 
+/// Determines the [ui.AppLifecycleState] by listening to 'focus', 'blur', and
+/// 'beforeunload' events on the [DomWindow].
+class AppLifecycleStateSupport {
```

Comment:
This implementation assumes we have a single global `FlutterView` that owns the entire page. Coincidentally, we are now introducing multi-view support where one "app" can have multiple `FlutterView` objects and the page's lifecycle may be owned by the host page/app, which may not even be written in Flutter or Dart (e.g. the host app can be a ReactJS app). The full-page mode isn't going away, however, so this functionality is reasonable to have. But maybe @ditman has some recommendations about how to add it in a way that won't make it more difficult to introduce the multi-view mode.

Reply:
The biggest concern I have with this implementation is that the `focus`/`blur` events on window will trigger more often than expected. For example, when opening a "file selector", it is common practice to rely on window focus/blur to detect the moment that users selects (or doesn't) a file, or when opening another browser window (while the current one is still visible) probably shouldn't stop the execution of the current flutter app.

I'd recommend using the Page Visibility API[^1] (`onvisibilitychange` and `visibilityState`) rather than `onfocus`/`onblur`.

From a multi-view perspective, we might want to expose some JS API for host pages to programmatically control the lifecycle of the embedded Flutter App, (so it can be torn down properly when a ReactJS component is unmounting, for example), but as a default implementation, the global `visibilitychange` and `beforeunload` events are so low level that should be enough in the most common cases.

[^1]: https://developer.mozilla.org/en-US/docs/Web/API/Page_Visibility_API